File: src/main/scala/category/timeline/week42/BSTIterator.scala
package category.timeline.week42

import category.timeline.week29.TreeNode


class BSTIterator(_root: TreeNode) {
  
  import scala.collection.mutable.ArrayBuffer
  
  val iterator: Iterator[Int] = {
    val buffer = new ArrayBuffer[Int]()
    
    def traversal(node: TreeNode): Unit = {
      if (node == null) return
      if (node.left == null && node.right == null) {
        buffer.append(node.value)
        return
      }
      traversal(node.left)
      buffer.append(node.value)
      traversal(node.right)
    }
    
    traversal(_root)
    buffer.iterator
  }
  
  /** @return the next smallest number */
  def next(): Int = {
    iterator.next()
  }
  
  /** @return whether we have a next smallest number */
  def hasNext(): Boolean = {
    iterator.hasNext
  }
  
  
}

File: src/main/python/medium/_150_200/BSTIterator.py
from src.main.python.TreeNode import TreeNode


class BSTIterator:

    def __init__(self, root: TreeNode):
        self.node = root
        self.stack = []
        while root: self.stack.append(root);root = root.left

    def next(self) -> int:
        """
        @return the next smallest number
        """
        ans = self.stack.pop()
        value = ans.val
        ans = ans.right
        while ans: self.stack.append(ans);ans = ans.left
        return value

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return self.stack
